### PR TITLE
Multiple remote reactive power control generators

### PR DESCRIPTION
**Pull request by caioluke**
**Please check if the PR fulfills these requirements**

- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem?**

**What kind of change does this PR introduce?**

Feature

**What is the current behavior?**

Only one generator can remotely control the reactive power of a branch.

**What is the new behavior (if this is a feature change)?**
It will allow multiple generators to remotely control the reactive power of a branch.
The generators can be on the same bus or in different buses.

**Does this PR introduce a breaking change or deprecate an API?**

No

**Other information**:

n/a

### CODE REVIEW

File: src/main/java/com/powsybl/openloadflow/ac/equations/AcEquationType.java [hunk 1]
```diff
@@ -22,7 +22,8 @@ public enum AcEquationType implements Quantity {
     BRANCH_TARGET_Q("branch_target_q", ElementType.BRANCH), // generator reactive power control
     BRANCH_TARGET_ALPHA1("branch_target_\u03B1", ElementType.BRANCH), // phase shifter constant shift
     BRANCH_TARGET_RHO1("branch_target_\u03C1", ElementType.BRANCH), // transformer constant voltage control
-    DISTR_Q("distr_q", ElementType.BUS), // remote voltage control reactive power distribution
+    DISTR_Q_VOLTAGE_CONTROL("distr_q_voltage_control", ElementType.BUS), // remote voltage control reactive power distribution
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/ac/equations/AcEquationType.java:25]: You have doubled all the code but the logic is very similar between distribution equations for voltage control and for reactive power remote control. I am quite convinced that everything can be mutualized with a deep understanding of both feature.

File: src/main/java/com/powsybl/openloadflow/ac/nr/PerEquationTypeStoppingCriteria.java [hunk 2]
```diff
@@ -61,7 +61,7 @@ private boolean computeStop(double[] fx, EquationSystem<AcVariableType, AcEquati
                     break;
                 case BRANCH_TARGET_Q,
                      BUS_TARGET_Q,
```
Comment (Hadrien-Godard) [inline src/main/java/com/powsybl/openloadflow/ac/nr/PerEquationTypeStoppingCriteria.java:63]: Why this indentation?

**Review by Hadrien-Godard — APPROVED** [review]

File: src/main/java/com/powsybl/openloadflow/ac/nr/PerEquationTypeStoppingCriteria.java (hunk 2 above)
Comment (caioluke) [reply src/main/java/com/powsybl/openloadflow/ac/nr/PerEquationTypeStoppingCriteria.java:63]: Missclick

File: src/test/java/com/powsybl/openloadflow/sa/OpenSecurityAnalysisTest.java [hunk 3]
```diff
@@ -1925,8 +1925,46 @@ void testWithVoltageRemoteControl() {
...
+        LoadFlowParameters lfParameters = new LoadFlowParameters();
+        lfParameters.setUseReactiveLimits(true);
+        OpenLoadFlowParameters openLoadFlowParameters = new OpenLoadFlowParameters();
+        openLoadFlowParameters.setReactivePowerRemoteControl(true);
+        lfParameters.addExtension(OpenLoadFlowParameters.class, openLoadFlowParameters);
+        SecurityAnalysisParameters securityAnalysisParameters = new SecurityAnalysisParameters();
+        securityAnalysisParameters.setLoadFlowParameters(lfParameters);
+        assertDoesNotThrow(() -> {
+            runSecurityAnalysis(network, contingencies, Collections.emptyList(), securityAnalysisParameters);
+        });
```
Comment (jeandemanged) [inline src/test/java/com/powsybl/openloadflow/sa/OpenSecurityAnalysisTest.java:1940]: should check results too

File: src/test/java/com/powsybl/openloadflow/sa/OpenSecurityAnalysisTest.java [hunk 4]
```diff
@@ -1925,8 +1925,46 @@ void testWithVoltageRemoteControl() {
...
```
Comment (jeandemanged) [inline src/test/java/com/powsybl/openloadflow/sa/OpenSecurityAnalysisTest.java:1958]: should check results too

**Review by jeandemanged — CHANGES_REQUESTED** [review]

File: src/main/java/com/powsybl/openloadflow/ac/equations/AcEquationSystemCreator.java [hunk 5]
```diff
@@ -216,7 +197,70 @@ private void createGeneratorRemoteVoltageControlEquations(GeneratorVoltageContro
         }
 
         // create reactive power distribution equations at voltage controller buses
-        createReactivePowerDistributionEquations(voltageControl, equationSystem, creationParameters);
+        createReactivePowerDistributionEquations(voltageControl.getMergedControllerElements(), equationSystem, creationParameters);
+    }
+
+    private void createReactivePowerControlEquations(ReactivePowerControl reactivePowerControl, EquationSystem<AcVariableType, AcEquationType> equationSystem) {
+        for (LfBus controllerBus : reactivePowerControl.getControllerBuses()) {
+            equationSystem.createEquation(controllerBus, AcEquationType.BUS_TARGET_Q);
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/ac/equations/AcEquationSystemCreator.java:205]: @geofjamg I think we don't have to do that, right?

Comment (annetill) [reply src/main/java/com/powsybl/openloadflow/ac/equations/AcEquationSystemCreator.java:205]: See #900

File: src/test/java/com/powsybl/openloadflow/ac/GeneratorRemoteControlTest.java [hunk 6]
```diff
@@ -476,6 +476,134 @@ void testSharedRemoteReactivePowerControl() {
...
+                .withEnabled(true).add();
+        g4n1.newExtension(RemoteReactivePowerControlAdder.class)
+                .withTargetQ(targetQ)
+                .withRegulatingTerminal(l34n1.getTerminal(Branch.Side.TWO))
+                .withEnabled(true).add();
+
+        parameters.getExtension(OpenLoadFlowParameters.class).setReactivePowerRemoteControl(true);
+        LoadFlowResult result1 = loadFlowRunner.run(network1, parameters);
+        assertTrue(result1.isOk());
+        assertEquals(targetQ, l34n1.getTerminal(Branch.Side.TWO).getQ(), 1E-2); // lower tolerance
```
Comment (annetill) [inline src/test/java/com/powsybl/openloadflow/ac/GeneratorRemoteControlTest.java:502]: I really wonder why our tolerance is so "bad"...

File: src/main/java/com/powsybl/openloadflow/network/LfContingency.java [hunk 7]
```diff
@@ -152,6 +152,7 @@ public void apply(LoadFlowParameters.BalanceType balanceType) {
             generator.setDisabled(true);
             if (generator.getGeneratorControlType() != LfGenerator.GeneratorControlType.OFF) {
                 generator.setGeneratorControlType(LfGenerator.GeneratorControlType.OFF);
+                bus.getReactivePowerControl().ifPresent(rc -> rc.updateReactiveKeys());
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/network/LfContingency.java:155]: @geofjamg I think this is missing for voltage control. @caioluke is going to publish an issue.

Comment (caioluke) [reply src/main/java/com/powsybl/openloadflow/network/LfContingency.java:155]: I just opened a new PR with the fix: https://github.com/powsybl/powsybl-open-loadflow/pull/904

**Review by annetill — APPROVED** [review]

**Review by jeandemanged — APPROVED** [review]
❤️

**Pull request merged**